src/uct/api/uct.h

```diff
@@ -2082,6 +2083,37 @@ ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
                               unsigned flags, const char *name, uct_mem_h *memh_p);
 
 
+/**
+ * @ingroup UCT_MD
+ * @brief Allocate memory of a specified memory type for zero-copy sends and
+ * remote access.
+ *
+ * Allocate memory on the memory domain. In order to use this function, MD
+ * must support @ref UCT_MD_FLAG_ALLOC flag on the specified memory type.
+ *
+ * @param [in]     md          Memory domain to allocate memory on.
+ * @param [in,out] length_p    Points to the size of memory to allocate. Upon successful
+ *                             return, filled with the actual size that was allocated,
+ *                             which may be larger than the one requested. Must be >0.
+ * @param [in,out] address_p   The address
+ * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
+ *                             Note that if UCT_MD_MEM_FLAG_FIXED is provided
+ *                             and the underlying allocator is unable to
+ *                             gaurantee an allocation starting at the specified
+ *                             address, an appropritate error is returned.
+ * @param [in]     mem_type    Type of memory to be allocated.
+ *                             E.g. UCS_MEMORY_TYPE_HOST, UCS_MEMORY_TYPE_CUDA,
+ *                             UCS_MEMORY_TYPE_ROCM, etc.
+ * @param [in]     name        Name of the allocated region, used to track memory
+ *                             usage for debugging and profiling.
+ * @param [out]    memh_p      Filled with handle for allocated region.
+ */
+ucs_status_t uct_md_mem_alloc_mem_type(uct_md_h md, size_t *length_p,
```

Comment:
instead of adding new API, we should extend uct_md_mem_alloc() with params(and field_mask) struct, which will also take the memory type

Reply:
@yosefe @brminich Adding a draft of extending uct_md_mem_alloc/uct_mem_alloc borrowing ucp_request_param_t additions in UCP. LMK if this is along the lines we should change
```diff
+/**
+ * @ingroup UCT_MD
+ * @brief UCT allocation parameter specification field
+ *
+ * The enumeration allows specifying which fields in @ref uct_mem_alloc_param_t
+ * are present.
+ */
+typedef enum {
+    UCT_ALLOC_ATTR_FIELD_FLAGS    = UCS_BIT(0) /**< Allocation flags field*/
+    UCT_ALLOC_ATTR_FIELD_MEM_TYPE = UCS_BIT(1) /**< Memory type field */
+} uct_alloc_attr_t;
+
+
+/**
+ * @ingroup UCT_MD
+ * @brief Allocation characteristics parameter passed to @ref uct_mem_alloc,
+ *        @ref uct_md_mem_alloc
+ *
+ * The structure @ref uct_mem_alloc_param_t is used to specify the memory type
+ * of the allocation, and flags indicating other characteristics of the
+ * allocation.
+ */
+typedef struct {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_alloc_attr_t. Fields not specified in this mask will be
+     * ignored
+     */
+    uint32_t       alloc_attr_mask;
+
+    /**
+     * Type of memory to be allocated.
+     * E.g. UCS_MEMORY_TYPE_HOST, UCS_MEMORY_TYPE_CUDA, UCS_MEMORY_TYPE_ROCM.
+     */
+    ucs_memory_type_t mem_type;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned flags;
+
+} uct_mem_alloc_param_t;
+
+
 /**
  * @ingroup UCT_MD
  * @brief Allocate memory for zero-copy sends and remote access.
@@ -2073,13 +2118,15 @@ ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr);
  *                             return, filled with the actual size that was allocated,
  *                             which may be larger than the one requested. Must be >0.
  * @param [in,out] address_p   The address
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_param_t.
  * @param [in]     name        Name of the allocated region, used to track memory
  *                             usage for debugging and profiling.
  * @param [out]    memh_p      Filled with handle for allocated region.
  */
 ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *name, uct_mem_h *memh_p);
+                              uct_mem_alloc_param_t *param, const char *name,
+                              uct_mem_h *memh_p);
 
 
 /**
@@ -2180,7 +2227,8 @@ ucs_status_t uct_md_detect_memory_type(uct_md_h md, const void *addr,
  *                             cannot be made, the allocation request fails.
  * @param [in]     min_length  Minimal size to allocate. The actual size may be
  *                             larger, for example because of alignment restrictions.
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_param_t.
  * @param [in]     methods     Array of memory allocation methods to attempt.
  * @param [in]     num_methods Length of 'methods' array.
  * @param [in]     mds         Array of memory domains to attempt to allocate
@@ -2192,7 +2240,8 @@ ucs_status_t uct_md_detect_memory_type(uct_md_h md, const void *addr,
  * @param [out]    mem         In case of success, filled with information about
  *                              the allocated memory. @ref uct_allocated_memory_t.
  */
-ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
+ucs_status_t uct_mem_alloc(void *addr, size_t min_length,
+                           uct_mem_alloc_param_t *param,
                            uct_alloc_method_t *methods, unsigned num_methods,
                            uct_md_h *mds, unsigned num_mds, const char *name,
                            uct_allocated_memory_t *mem);

```

Reply:
looks good to me. Maybe also move `mds` and `num_mds` of `uct_mem_alloc` to params, because they are optional?